### PR TITLE
Issue 59: Promote pg_worker to first-class binary with PATH autodiscovery (Vibe Kanban)

### LINKED ISSUE #59: Make pg_worker a first‑class binary + document root usage
## Problem

`pg_worker` is essential for the library’s core root-mode workflow, but it’s currently
treated like a test helper:

- It lives under `tests/support/pg_worker.rs`.
- It’s gated behind the `dev-worker` feature.
- `cargo install --path . installs only pg_embedded_setup_unpriv`, not the worker.
- Documentation doesn’t clearly explain how to build/install `pg_worker` or how root users
  should configure `PG_EMBEDDED_WORKER`.

This creates friction for library consumers and makes the root flow feel “hidden” or
unreliable.

## Impact

- Root-mode users have to discover and build the worker manually.
- CI and local usage require non-obvious --features dev-worker flags.
- The worker is not exposed as a stable, installable binary even though it is required for
  core functionality.

## Proposed changes

1. Promote pg_worker to a first‑class binary:
    - Move to src/bin/pg_worker.rs.
    - Remove dev-worker feature gate for the worker binary.
2. Make installation straightforward:
    - cargo install --path . should install both pg_embedded_setup_unpriv and pg_worker
      (or document explicit --bin pg_worker).
3. Autodiscover the worker if it on the envrionment's binary search path.
4. Improve docs:
    - Add a “Root usage” section in README.md and docs/users-guide.md.
    - Clearly state that `PG_EMBEDDED_WORKER` must point to `pg_worker` if a custom worker is needed.
    - Provide example install/build commands.

## Acceptance criteria

- pg_worker is built by default (no feature gate).
- Documentation explicitly describes how to install and configure `PG_EMBEDDED_WORKER`.
- Root usage is easy to follow with a minimal copy/paste example.

## Notes

This is core functionality, not just test scaffolding. Treating the worker as a test helper contradicts the library’s main purpose for root users.

### PR DESCRIPTION
**Pull request by leynos**
## Summary

Promotes `pg_worker` from a test helper to a first-class binary that installs by default via `cargo install`, and adds PATH-based autodiscovery so root users don't need explicit configuration.

### Binary Promotion
- Moved `pg_worker` from `tests/support/pg_worker.rs` to `src/bin/pg_worker.rs`
- Removed `required-features = ["dev-worker"]` gate from `Cargo.toml`
- `cargo install pg-embed-setup-unpriv` now installs both `pg_embedded_setup_unpriv` and `pg_worker`

### PATH-based Autodiscovery
- Added `discover_worker_from_path()` function to search PATH directories for `pg_worker`
- Discovery order: `PG_EMBEDDED_WORKER` env var first, then PATH search
- Zero-configuration usage when worker is installed to `~/.cargo/bin`

### Improved Error Messages
- Updated error message when worker binary is not found to guide users on installation and configuration options

### Documentation
- Added "Root usage and the worker binary" section to `README.md`
- Updated "Platform expectations" in `docs/users-guide.md` to mention autodiscovery
- Added installation instructions to "Integrating with root-only test agents" section

## Why

Previously, `pg_worker` was treated as a test helper despite being essential for the library's core root-mode workflow:
- It was gated behind the `dev-worker` feature
- `cargo install` only installed `pg_embedded_setup_unpriv`, not the worker
- Documentation didn't clearly explain how to build/install or configure `PG_EMBEDDED_WORKER`

This created friction for library consumers and made the root flow feel hidden or unreliable.

## Test Plan

- [x] All 167 tests pass (1 pre-existing timeout, 2 skipped as expected)
- [x] Lint checks pass (`make lint`)
- [x] Format checks pass (`make check-fmt`)
- [x] Markdown lint passes (`make markdownlint`)

Fixes #59

---

*This PR was written using [Vibe Kanban](https://vibekanban.com)*

## Summary by Sourcery

Promote the pg_worker helper to a first-class installed binary and add PATH-based autodiscovery with clearer root-mode worker configuration errors and docs.

New Features:
- Add PATH-based autodiscovery of the pg_worker binary with a defined precedence over the PG_EMBEDDED_WORKER environment variable.
- Promote pg_worker from a test-only helper to a normal installed binary in the crate.

Enhancements:
- Improve validation around the configured worker binary path, including checks for empty or root paths and UTF-8 handling.
- Update root-mode bootstrap error messaging to guide users on installing or configuring the pg_worker binary.
- Adjust bootstrap tests to account for optional worker discovery even in unprivileged mode.

Documentation:
- Document root-mode usage, installation, and discovery of the pg_worker binary in the README and user guide, including updated guidance for root-only test agents.

Tests:
- Update tests to reflect the new worker autodiscovery behavior and error messages when the worker binary is missing.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[src/bootstrap/env.rs](https://github.com/leynos/pg-embedded-setup-unpriv/pull/68/files/613ae15c90149b9d9bff1e6df2527c26f810e7a8#diff-9faebad3a459dc2a666adbcdfb2b96301e3e6fc7c97850b348ab2ae186305bda)

Comment on lines +478 to +502

~~~rust
    fn discover_worker_skips_directories() {
        let temp = tempdir().expect("create tempdir");
        let worker_dir = temp.path().join("pg_worker");
        fs::create_dir(&worker_dir).expect("create directory");

        let original_path = std::env::var_os("PATH");
        let new_path = temp.path().to_string_lossy().to_string();
        unsafe {
            // SAFETY: test is single-threaded for this env modification
            std::env::set_var("PATH", &new_path);
        }

        let result = discover_worker_from_path();

        // Restore PATH
        match original_path {
            Some(p) => unsafe { std::env::set_var("PATH", p) },
            None => unsafe { std::env::remove_var("PATH") },
        }

        assert!(
            result.is_none(),
            "should not find pg_worker when it is a directory"
        );
    }
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/72457/delta?repo-id=1225432&review-id=68&biomarker=Code+Duplication&filename=src%2Fbootstrap%2Fenv.rs&method=)
The module contains 2 functions with similar structure: tests.discover_worker_returns_none_when_not_found,tests.discover_worker_skips_directories

File: src/bootstrap/env.rs [hunk 1]
```diff
@@ -123,9 +141,36 @@ pub(super) fn worker_binary_from_env() -> BootstrapResult<Option<Utf8PathBuf>> {
...
+    #[cfg(windows)]
+    let candidate = candidate.with_extension("exe");
+    candidate
+}
+
+fn is_valid_worker_candidate(candidate: &std::path::Path) -> Option<Utf8PathBuf> {
+    if !candidate.is_file() {
+        return None;
+    }
+    Utf8PathBuf::from_path_buf(candidate.to_path_buf()).ok()
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/bootstrap/env.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

~~~
Please address the comments from this code review:

## Overall Comments
- In `discover_worker_from_path`/`is_valid_worker_candidate`, consider returning a more specific error when PATH contains a non-UTF-8 directory or candidate instead of silently skipping it, so users can diagnose misconfigured environments more easily.
- The PATH search currently hardcodes `pg_worker` plus `.exe` on Windows; if you ever introduce a configurable worker name or additional suffixes (e.g., debug builds), it might be worth centralizing the worker binary name/extension logic to avoid duplication.
- The new helper functions in `env.rs` (`try_find_worker_in_dir`, `build_worker_candidate_path`, `is_valid_worker_candidate`) are quite small and tightly coupled; consider inlining or reducing their number to keep the bootstrap PATH discovery code easier to follow.

## Individual Comments

### Comment 1
<location> `src/bootstrap/env.rs:169-173` </location>
<code_context>
+    candidate
+}
+
+fn is_valid_worker_candidate(candidate: &std::path::Path) -> Option<Utf8PathBuf> {
+    if !candidate.is_file() {
+        return None;
+    }
+    Utf8PathBuf::from_path_buf(candidate.to_path_buf()).ok()
 }

</code_context>

<issue_to_address>
**suggestion (performance):** Avoid the extra `PathBuf` allocation when converting a candidate path to `Utf8PathBuf`.

`build_worker_candidate_path` already returns a `PathBuf`, but `is_valid_worker_candidate` takes `&Path` and calls `to_path_buf()`, causing an extra allocation. You can eliminate this by taking ownership of the `PathBuf` instead:

```rust
fn try_find_worker_in_dir(dir: &std::path::Path) -> Option<Utf8PathBuf> {
    let candidate = build_worker_candidate_path(dir);
    is_valid_worker_candidate(candidate)
}

fn is_valid_worker_candidate(candidate: PathBuf) -> Option<Utf8PathBuf> {
    if !candidate.is_file() {
        return None;
    }
    Utf8PathBuf::from_path_buf(candidate).ok()
}
```
</issue_to_address>

### Comment 2
<location> `src/bootstrap/env.rs:103` </location>
<code_context>
 }

 pub(super) fn worker_binary_from_env() -> BootstrapResult<Option<Utf8PathBuf>> {
-    let Some(raw) = env::var_os("PG_EMBEDDED_WORKER") else {
-        return Ok(None);
</code_context>

<issue_to_address>
**issue (complexity):** Consider inlining the single-use helper functions so the worker discovery and env parsing logic flows linearly within fewer functions.

You can keep all the new behavior while reducing indirection by inlining some of the tiny helpers that are only used once and form a linear flow.

### 1. Flatten the PATH discovery chain

`discover_worker_from_path → try_find_worker_in_dir → build_worker_candidate_path → is_valid_worker_candidate` is effectively a single linear operation. You can keep the semantics but express it directly:

```rust
fn discover_worker_from_path() -> Option<Utf8PathBuf> {
    let path_var = env::var_os("PATH")?;
    env::split_paths(&path_var)
        .map(|dir| {
            let candidate = dir.join("pg_worker");
            #[cfg(windows)]
            let candidate = candidate.with_extension("exe");
            candidate
        })
        .find(|candidate| candidate.is_file())
        .and_then(|candidate| Utf8PathBuf::from_path_buf(candidate).ok())
}
```

This removes three helpers and keeps all the logic in one place, making the control flow easier to follow while preserving:
- scanning `PATH`
- platform-specific `.exe` handling
- UTF-8-only paths (non-UTF-8 are skipped via `.ok()`).

### 2. Merge `parse_worker_path_from_env` and `validate_worker_path_not_empty_or_root`

The validation is tightly coupled to parsing and only used once. You can fold `validate_worker_path_not_empty_or_root` into `parse_worker_path_from_env` and drop the extra function:

```rust
fn parse_worker_path_from_env(raw: &std::ffi::OsStr) -> BootstrapResult<Utf8PathBuf> {
    let path = Utf8PathBuf::from_path_buf(PathBuf::from(raw)).map_err(|_| {
        let invalid_value = raw.to_string_lossy().to_string();
        BootstrapError::from(color_eyre::eyre::eyre!(
            "PG_EMBEDDED_WORKER contains a non-UTF-8 value: {invalid_value:?}. \
             Provide a UTF-8 encoded absolute path to the worker binary."
        ))
    })?;

    if path.as_str().is_empty() {
        return Err(BootstrapError::from(color_eyre::eyre::eyre!(
            "PG_EMBEDDED_WORKER must not be empty"
        )));
    }
    if path.as_str() == "/" {
        return Err(BootstrapError::from(color_eyre::eyre::eyre!(
            "PG_EMBEDDED_WORKER must not point at the filesystem root"
        )));
    }

    Ok(path)
}
```

`worker_binary_from_env` stays the same in terms of behavior, but the number of one-use helpers drops, reducing cognitive load without changing functionality.
</issue_to_address>

### Comment 3
<location> `src/bootstrap/env.rs:103` </location>
<code_context>
 }

 pub(super) fn worker_binary_from_env() -> BootstrapResult<Option<Utf8PathBuf>> {
-    let Some(raw) = env::var_os("PG_EMBEDDED_WORKER") else {
-        return Ok(None);
</code_context>

<issue_to_address>
**issue (review_instructions):** Add behavioural tests that exercise the new PATH‑based worker discovery behaviour in `worker_binary_from_env`.

You changed `worker_binary_from_env` to fall back to `discover_worker_from_path()` when `PG_EMBEDDED_WORKER` is not set, but there is no new behavioural test in this diff that verifies the end‑to‑end behaviour for this case (e.g. worker discovered via PATH vs. not found). Add an integration/behavioural test that sets up PATH with and without a `pg_worker` binary and asserts the resulting execution mode / error paths so this new behaviour is covered.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*`

**Instructions:**
For any new feature or change to an existing feature, both behavioural *and* unit tests are required.

</details>
</issue_to_address>

### Comment 4
<location> `src/bootstrap/env.rs:152` </location>
<code_context>
+/// Returns the first valid UTF-8 path to an existing `pg_worker` file, or
+/// `None` if no candidate is found. This enables zero-configuration usage when
+/// the worker is installed via `cargo install`.
+fn discover_worker_from_path() -> Option<Utf8PathBuf> {
+    let path_var = env::var_os("PATH")?;
+    env::split_paths(&path_var).find_map(|dir| try_find_worker_in_dir(&dir))
</code_context>

<issue_to_address>
**issue (review_instructions):** Add unit tests for `discover_worker_from_path` and its helpers to validate PATH scanning logic and edge cases.

You introduced `discover_worker_from_path`, `try_find_worker_in_dir`, `build_worker_candidate_path`, and `is_valid_worker_candidate`, but there are no corresponding unit tests in this diff. Add unit tests that cover at least:
- a PATH entry containing a valid `pg_worker` file (and `.exe` on Windows),
- PATH entries with no worker file,
- non‑UTF‑8 PATH segments and invalid UTF‑8 candidate paths,
- directories where a `pg_worker` path exists but is not a regular file.
This will satisfy the requirement for unit tests on new logic and guard against regressions in PATH‑based discovery.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*`

**Instructions:**
For any new feature or change to an existing feature, both behavioural *and* unit tests are required.

</details>
</issue_to_address>

### Comment 5
<location> `src/bootstrap/env.rs:120` </location>
<code_context>
+    Ok(None)
+}
+
+fn parse_worker_path_from_env(raw: &std::ffi::OsStr) -> BootstrapResult<Utf8PathBuf> {
+    let path = Utf8PathBuf::from_path_buf(PathBuf::from(raw)).map_err(|_| {
         let invalid_value = raw.to_string_lossy().to_string();
</code_context>

<issue_to_address>
**issue (review_instructions):** Add unit tests for `parse_worker_path_from_env` and `validate_worker_path_not_empty_or_root` to cover new validation behaviour.

You factored out `parse_worker_path_from_env` and `validate_worker_path_not_empty_or_root` and added explicit error cases for empty strings and root paths, but there are no new unit tests exercising these functions. Add focused tests that verify:
- non‑UTF‑8 `PG_EMBEDDED_WORKER` values produce the documented error,
- empty strings and root (`/` or equivalent) paths are rejected with the expected messages,
- a valid non‑empty, non‑root path is accepted. 
This will ensure the new validation behaviour is properly specified by tests, as required.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*`

**Instructions:**
For any new feature or change to an existing feature, both behavioural *and* unit tests are required.

</details>
</issue_to_address>

### Comment 6
<location> `docs/users-guide.md:23` </location>
<code_context>
-  `PG_EMBEDDED_WORKER` so the helper can drop to `nobody` for filesystem work.
+- Linux supports both privilege branches. Root executions use `pg_worker` to
+  drop to `nobody` for filesystem work. The library discovers the worker
+  automatically if it is in `PATH`, or you can set `PG_EMBEDDED_WORKER`
+  explicitly.
 - macOS runs the unprivileged path; root executions are expected to fail fast
</code_context>

<issue_to_address>
**suggestion (review_instructions):** This sentence uses the second-person pronoun "you", which violates the no first/second person instruction.

Consider rephrasing to avoid directly addressing the reader. For example:

- "..., or `PG_EMBEDDED_WORKER` can be set explicitly."
- "..., or set `PG_EMBEDDED_WORKER` explicitly."

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Avoid 2nd person or 1st person pronouns ("I", "you", "we").

</details>
</issue_to_address>
~~~

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review

File: src/bootstrap/env.rs [hunk 2]
```diff
@@ -124,8 +139,103 @@ pub(super) fn worker_binary_from_env() -> BootstrapResult<Option<Utf8PathBuf>> {
...
+pub(crate) fn discover_worker_from_path(worker_name: &str) -> BootstrapResult<Option<Utf8PathBuf>> {
+    let path_var = match env::var_os("PATH") {
+        Some(v) => v,
+        None => return Ok(None),
+    };
+
+    let mut error = None;
+
+    for dir in env::split_paths(&path_var) {
+        let dir_str = match Utf8PathBuf::from_path_buf(dir.clone()) {
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/bootstrap/env.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/bin/pg_worker.rs [hunk 3]
```diff
@@ -1,6 +1,22 @@
...
+//!
+//! Install via Cargo:
+//!
+//! ```bash
+//! cargo install pg-embed-setup-unpriv
+//! ```
+//!
+//! This installs both `pg_embedded_setup_unpriv` (the main setup helper) and
+//! `pg_worker` (this binary). The library discovers `pg_worker` automatically
+//! if it is in your `PATH`, or you can set `PG_EMBEDDED_WORKER` explicitly.
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/bin/pg_worker.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/bootstrap/env_tests.rs [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+    Some("/usr/local/bin/pg_worker")
+)]
+#[cfg(unix)]
+#[case(
+    OsStr::from_bytes(b"/path/with/invalid/\xff/bytes"),
+    true,
+    "non-UTF-8 value",
+    None
+)]
+fn parse_worker_path_cases(
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/bootstrap/env_tests.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/bootstrap/worker_discovery.rs [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+    } else {
+        BootstrapError::from(report)
+    }
+}
+
+fn error_chain_has_not_found(err: &Report) -> bool {
+    err.chain()
+        .filter_map(|source| source.downcast_ref::<std::io::Error>())
+        .any(|source| source.kind() == ErrorKind::NotFound)
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/bootstrap/worker_discovery.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.
